orttraining/orttraining/core/agent/training_agent.cc

```diff
@@ -0,0 +1,136 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "orttraining/core/agent/training_agent.h"
+#include "core/session/IOBinding.h"
+#include "orttraining/training_ops/cpu/controlflow/ort_tasks.h"
+
+namespace onnxruntime {
+namespace training {
+
+TrainingAgent::TrainingAgent(InferenceSession *session) : inference_session_(session) {}
+
+TrainingAgent::~TrainingAgent() {
+  // TODO: Properly cancel outstanding background tasks
+  // Following implementation only handle the case where bg_thread is waiting for backward inputs
+  // Background thread can also be in other states, such as running Forward() or running Backward()
+  std::vector<int64_t> run_ids;
+  {
+    std::lock_guard<std::mutex> lock(bg_threads_mutex_);
+    for (auto it = bg_threads_.begin(); it != bg_threads_.end(); ++it) {
+      run_ids.push_back(it->first);
+    }
+  }
+  for (int64_t run_id : run_ids) {
+    if (!onnxruntime::contrib::OrtTasks::GetInstance().TaskIsCompleted(run_id)) {
+      CancelPendingBackwardRun(run_id);
+    }
+  }
+};
+
+common::Status TrainingAgent::RunForward(const RunOptions& run_options, onnxruntime::IOBinding& io_binding,
+                                         std::vector<OrtValue>& user_outputs, int64_t& run_id) {
+  std::promise<void> setup_promise;
+  std::future<void> setup_future = setup_promise.get_future();
+
+  // Passing run_options and io_binding by reference to the bg_thread,
+  // this is ok because they are ORTModule's member, and they are presistent through forward and backward calls
+  auto bg_thread = std::thread([this](std::future<void> setup_future, const RunOptions& run_options, onnxruntime::IOBinding& io_binding) {
+    // wait until task is properly setup
+    setup_future.get();
+
+    common::Status status = inference_session_->Run(run_options, io_binding);
+
+    onnxruntime::contrib::OrtTasks::GetInstance().SetStatus(status);
+
+    // If forward outputs still hasn't been consumed at this point, i.e. forward function hasn't complete itself
+    // this indicates that Run() call returned before hitting YieldOp, due to hitting some exception during the forward subgraph execution
+    // In this case, we need to wake up the foreground thread and pass along the failed status.
+    // Otherwise, foreground thread will be stuck waiting for forward_outputs.
+    if (onnxruntime::contrib::OrtTasks::GetInstance().ForwardOutputsIsValid()) {
+      ORT_ENFORCE(!status.IsOK());
+      // signal main thread for background thread completion
+      onnxruntime::contrib::OrtTasks::GetInstance().SetForwardOutputs(status, {});
+    }
+  }, std::move(setup_future), std::cref(run_options), std::ref(io_binding));
+
+  run_id = std::hash<std::thread::id>()(bg_thread.get_id());
```

Comment:
>run_id = std::hash<std::thread::id>()(bg_thread.get_id()); [](start = 2, length = 58)

why not use thread::id as the map key?

Reply:
we need to return the key to python, so we first converted the std::thread::id to int64_t. 
otherwise, we would need to pybind std::thread::id to some python type.